app/controllers/users/emails_controller.rb

```diff
@@ -15,6 +15,11 @@ def show
       @add_user_email_form = AddUserEmailForm.new
     end
 
+    def sp_show
```

Comment:
The ParseControllerFromReferrer seems to be working for this. Do you think the second problem above would still exist? 
> fail form submission
> cancel link now goes to the account dashboard

Reply:
I think it'd still be a problem? Worth testing, maybe including test coverage.

Reply:
If it ends up being a problem, we could try to make sure that we maintain some detail about where they came from, maybe with a hidden form input that contains the value of the parsed referer, so we know where to send them after or if they cancel.

e.g.

```rb
def show
  # ...
  @adding_from_completions_screen = adding_from_completions_screen?
end
```

```erb
<%= f.input(:adding_from_completions_screen, as: :hidden, input_html: { value: @adding_from_completions_screen }) %>
```

```rb
def add
  # ...
  if result.success?
    # ...
  else
    @adding_from_completions_screen = adding_from_completions_screen?
  end
end

# ...

def adding_from_completions_screen?
  parsed_referer[:request_came_from] == 'sign_up/completions#show' ||
    permitted_params[:adding_from_completions_screen] == 'true'
end
```

Reply:
It was especially problematic like if a user had to re-auth they'd lose any query string set. 

I have a user_session now being set and destroyed in the completions controller. Do you see any potential complications from that? If so I'd try what you're suggesting above. 

I think the session token should give a persisting value that the cancel link can use though.

Reply:
The only issue with user session value is that we're not always consistent about _cleaning up_ session values, and it's sometimes tricky to know when that clean-up should occur. This can create problems in the sense of user session being littered with lots of different values, and also if the user would visit this screen multiple times in the same session from different flows, the session value might cause unexpected results.

e.g.

1. Start from partner
2. Sign in
3. From consent, change email
4. Consent to share info
5. Go to account dashboard (in production, this would be the user visiting https://secure.login.gov )
6. Go to Connected Accounts
7. Click "Change" 
8. Click "Back"

What I'd fear could happen in the scenario above is that we'd try to send them back to the completions screen because we had set the session value earlier at Step 3.

Reply:
I have the user_session token deleted at step 4. There doesn't seem to be any problems plus that back link at step 8 isn't the same as the cancel link. Not perfect but if you go to the completions url you're redirected to the account page since you're no longer in the creation flow.

FWIW I tested with that step flow above and didn't see anything strange happen. I instead of changing clicked to add an email and on that screen the cancel link seemed to be going where it should: back to the account dashboard.

Reply:
Ahh, sorry, I wrote that scenario before actually looking at the code 🤦 Glad you were already on top of it!